File: elevator-system/go-without-multithreading/elevator_controller.go
package main

import (
	"fmt"
	"math"
)

type ElevatorController struct {
	Elevators []*Elevator
}

func NewElevatorController(numElevators int, capacity int) *ElevatorController {
	elevators := make([]*Elevator, numElevators)
	for i := 0; i < numElevators; i++ {
		elevators[i] = NewElevator(fmt.Sprintf("E%d", i+1), capacity)
	}
	return &ElevatorController{
		Elevators: elevators,
	}
}

func (ec *ElevatorController) RequestElevator(req *Request) error {
	var optimalElevator *Elevator
	minDistance := math.MaxInt32

	for _, el := range ec.Elevators {
		dist := el.GetDistanceIfAssigned(req)
		if dist < minDistance {
			minDistance = dist
			optimalElevator = el
		}
	}

	if optimalElevator == nil {
		return fmt.Errorf("all elevators are currently over capacity to handle %d passengers", req.Passengers)
	}

	fmt.Printf("[Dispatcher] 🟢 Routing (Src: %d -> Dest: %d, Pax: %d) into Optimal => %s\n", req.SourceFloor, req.DestinationFloor, req.Passengers, optimalElevator.ID)
	return optimalElevator.AddRequest(req)
}

func (ec *ElevatorController) TickAll() bool {
	isActive := false
	for _, el := range ec.Elevators {
		el.Tick()
		if el.CurrentDirection != IDLE {
			isActive = true
		}
	}
	return isActive
}

File: elevator-system/java/ElevatorController.java
import java.util.ArrayList;
import java.util.List;

public class ElevatorController {
    private final List<Elevator> elevators;

    public ElevatorController(int numElevators, int capacity) {
        elevators = new ArrayList<>();
        for (int i = 0; i < numElevators; i++) {
            elevators.add(new Elevator("E" + (i + 1), capacity));
        }
    }

    public void requestElevator(Request req) throws Exception {
        Elevator optimalElevator = null;
        int minDistance = Integer.MAX_VALUE;

        for (Elevator el : elevators) {
            int dist = el.getDistanceIfAssigned(req);
            if (dist < minDistance) {
                minDistance = dist;
                optimalElevator = el;
            }
        }

        if (optimalElevator == null) {
            throw new Exception("all elevators are currently over capacity to handle " + req.getPassengers() + " passengers");
        }

        System.out.printf("[Dispatcher] \uD83D\uDFE2 Routing (Src: %d -> Dest: %d, Pax: %d) into Optimal => %s\n",
                req.getSourceFloor(), req.getDestinationFloor(), req.getPassengers(), optimalElevator.getId());
        
        optimalElevator.addRequest(req);
    }

    public boolean tickAll() {
        boolean isActive = false;
        for (Elevator el : elevators) {
            el.tick();
            if (el.getCurrentDirection() != Direction.IDLE) {
                isActive = true;
            }
        }
        return isActive;
    }
}
